File: 06_Searching/src/Split_Array_Largest_Sum_410.java
public class Split_Array_Largest_Sum_410 {
    public static int splitArray(int[] nums, int k) {
        int start = 0;
        int end = 0;

        for (int num : nums) {
            start = Math.max(start,num);
            end += num;
        }

        while (start < end) {
            int mid = start + (end-start)/2;

            if (counter(nums,mid,k)) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }

        return start;
    }

    public static boolean counter (int[] arr, int maxCount, int maxAns) {
        int cCount = 0;
        int ans = 1;

        for (int num : arr) {
            cCount += num;

            if (cCount > maxCount) {
                ans++;
                if (ans > maxAns) {
                    return false;
                }
                cCount = num;
            }
        }

        return true;
    }


    public static void main(String[] args) {
        System.out.println(splitArray(new int[] {7,2,5,10,8},2));
    }
}

File: 06_Searching/src/Split_Array_Largest_Sum_410.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var splitArray = function(nums, k) {
    let start = 0;
    let end = 0;

    for (let num of nums) {
        start = Math.max(start,num);
        end += num;
    }

    while (start < end) {
        let mid = Math.floor(start + (end-start)/2);

        if (counter(nums,mid,k)) {
            end = mid;
        } else {
            start = mid + 1;
        }
    }

    return start;
};

function counter (arr, maxSum, k) {
    let cSum = 0;
    let ans = 1;

    for (let num of arr) {
        cSum += num;

        if (cSum > maxSum) {
            ans++;
            if (ans > k) {
                return false;
            }
            cSum = num
        }
    }

    return true;
}

